br/pkg/lightning/restore/precheck_impl.go

```diff
@@ -0,0 +1,1164 @@
+			"For details of this failure, read the log file from the PREVIOUS run",
+			uniqueName, failedStep.MetricName(), action.String()))
+		return msgs, nil
+	}
+
+	dbInfos, err := ci.preInfoGetter.GetAllTableStructures(ctx)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	dbInfo, ok := dbInfos[tableInfo.DB]
+	if ok {
+		t, ok := dbInfo.Tables[tableInfo.Name]
+		if ok {
+			if tableCheckPoint.TableID > 0 && tableCheckPoint.TableID != t.ID {
+				msgs = append(msgs, fmt.Sprintf("TiDB Lightning has detected tables with illegal checkpoints. To prevent data loss, this run will stop now,"+
+					"please run command \"./tidb-lightning-ctl --checkpoint-remove='%s' --config=...\""+
+					"You may also run `./tidb-lightning-ctl --checkpoint-error-destroy=all --config=...` to start from scratch,"+
+					"For details of this failure, read the log file from the PREVIOUS run",
+					uniqueName))
+				return msgs, nil
+			}
+		}
+	}
+
+	var permFromCheckpoint []int
+	var columns []string
+	for _, eng := range tableCheckPoint.Engines {
+		if len(eng.Chunks) > 0 {
+			chunk := eng.Chunks[0]
+			permFromCheckpoint = chunk.ColumnPermutation
+			columns = chunk.Chunk.Columns
+			if filepath.Dir(chunk.FileMeta.Path) != ci.cfg.Mydumper.SourceDir {
+				message := fmt.Sprintf("chunk checkpoints path is not equal to config"+
+					"checkpoint is %s, config source dir is %s", chunk.FileMeta.Path, ci.cfg.Mydumper.SourceDir)
+				msgs = append(msgs, message)
+			}
+		}
+	}
+	if len(columns) == 0 {
+		log.FromContext(ctx).Debug("no valid checkpoint detected", zap.String("table", uniqueName))
+		return nil, nil
+	}
+	info := dbInfos[tableInfo.DB].Tables[tableInfo.Name]
+	if info != nil {
+		permFromTiDB, err := parseColumnPermutations(info.Core, columns, nil, log.FromContext(ctx))
+		if err != nil {
+			msgs = append(msgs, fmt.Sprintf("failed to calculate columns %s, table %s's info has changed,"+
+				"consider remove this checkpoint, and start import again.", err.Error(), uniqueName))
+		}
+		if !reflect.DeepEqual(permFromCheckpoint, permFromTiDB) {
+			msgs = append(msgs, fmt.Sprintf("compare columns perm failed. table %s's info has changed,"+
+				"consider remove this checkpoint, and start import again.", uniqueName))
+		}
+	}
+	return msgs, nil
+}
+
+type schemaCheckItem struct {
+	cfg           *config.Config
+	preInfoGetter PreRestoreInfoGetter
+	dbMetas       []*mydump.MDDatabaseMeta
+	checkpointsDB checkpoints.DB
+}
+
+func NewSchemaCheckItem(cfg *config.Config, preInfoGetter PreRestoreInfoGetter, dbMetas []*mydump.MDDatabaseMeta, cpdb checkpoints.DB) PrecheckItem {
+	return &schemaCheckItem{
+		cfg:           cfg,
+		preInfoGetter: preInfoGetter,
+		dbMetas:       dbMetas,
+		checkpointsDB: cpdb,
+	}
+}
+
+func (ci *schemaCheckItem) GetCheckItemID() CheckItemID {
+	return CheckSourceSchemaValid
+}
+
+func (ci *schemaCheckItem) Check(ctx context.Context) (*CheckResult, error) {
```

Comment:
This is actually the main loop in `(rc *Controller).DataCheck()` in `br/pkg/lightning/restore/restore.go`, L2072 - L2093, but each iteration only runs `SchemaIsValid()`